The parser crashes if no schema type is specified in the OpenAPI document

**Describe the bug**

The OpenAPI 3 specification prescribes that when no `type` is specified for a schema, this implicitly is equivalent to `anyOf` with all available types (see https://swagger.io/docs/specification/data-models/data-types/#any).

**To Reproduce**
Steps to reproduce the behavior:
1. Create an OpenAPI document with a schema with the following definition:
    ```yaml
    SomeSchema:
      description: Can be any value
    ```
2. Try to parse this schema
3. Observe a `KeyError` due to the missing `type` key

**Expected behavior**

The parser does not crash and correctly infers the `AnyOf` data type.

**System details (please complete the following information):**
 - OS: Ubuntu
 - OpenAPI 3
 - Python 3.10

**Additional context**

Mostly reporting the findings of https://github.com/maykinmedia/openapi3-parser/pull/1/ to upstream.

Looking at https://spec.openapis.org/oas/v3.0.3#schema-object (the same as your github link), it indeed states that `type` must be a string, but it doesn't specify that `type` is a required field/key/property (looking at `allOf` below that there is similar phrasing, but this property is a bit easier to reason about that there's no guarantee it's present).

The properties are taken from (and adjusted) the JSON Schema definition, so looking at that specification (https://datatracker.ietf.org/doc/html/draft-wright-json-schema-validation-00)  I can't find any rule saying that the validation keyword `type` must be present. More links on https://json-schema.org/specification-links.html

We are running into this with a particular schema (https://app.swaggerhub.com/apis/DH-Sandbox/handelsregister/1.3.0). Validating this with Spectral reports some errors but they are only related to the examples provided:

```
➜  validation ./node_modules/.bin/spectral lint /tmp/spec.yaml                                                                           
OpenAPI 3.x detected

/tmp/spec.yaml
 389:33  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.MaatschappelijkeActiviteitRaadplegen.allOf[1].properties.manifesteertZichAls
 391:31  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.MaatschappelijkeActiviteitRaadplegen.allOf[1].properties.wordtGeleidVanuit
 395:21  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.MaatschappelijkeActiviteitRaadplegen.allOf[1].properties.wordtUitgeoefendIn.items
 471:17  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.Onderneming.properties.wordtUitgeoefendIn.items
 596:27  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.NietNatuurlijkPersoonLijst.allOf[1].properties.bezoekLocatie
 610:21  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.NietNatuurlijkPersoonCollection.properties._embedded.properties.nietnatuurlijkpersonen.items
 624:27  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.NietNatuurlijkPersoon.allOf[1].properties.bezoekLocatie
 632:11  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.NietNatuurlijkPersoonRaadplegen.allOf[0]
 718:20  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.AdresBasis.properties.huisnummer.example
 747:11  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.NNPAdresRaadplegen.allOf[0]
 759:11  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.PostadresRaadplegen.allOf[0]
 779:11  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.AdresMetRegistratie.allOf[0]
 789:11  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.PostadresMetRegistratie.allOf[0]
 807:21  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingCollection.properties._embedded.properties.vestigingen.items
 832:27  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingLijst.allOf[1].properties.bezoekLocatie
 834:25  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingLijst.allOf[1].properties.postLocatie
 841:27  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingBeperkt.allOf[1].properties.bezoekLocatie
 843:25  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingBeperkt.allOf[1].properties.postLocatie
 859:27  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingRaadplegen.allOf[1].properties.bezoekLocatie
 861:25  error  oas3-valid-schema-example  `example` property type should be integer  components.schemas.VestigingRaadplegen.allOf[1].properties.postLocatie

✖ 20 problems (20 errors, 0 warnings, 0 infos, 0 hints)
```

One crashing example seems to be https://app.swaggerhub.com/apis/DH-Sandbox/handelsregister/1.3.0#/Geboorte - however looking at the rest of the schema I do believe they should include a `type: object` here, otherwise specifying `properties` seems to make little sense to me?

--- a/src/openapi_parser/builders/schema.py
+++ b/src/openapi_parser/builders/schema.py
@@ -47,7 +47,7 @@ def extract_attrs(data: dict, attrs_map: Dict[str, PropertyMeta]) -> Dict[str, A
         if data.get(name) is not None
     }
 
-    attrs['type'] = DataType(attrs['type'])
+    attrs['type'] = DataType(attrs.get('type', DataType.ANY_OF))
 
     attrs.update(extract_typed_props(data, attrs_map))
 
@@ -111,7 +111,7 @@ class SchemaFactory:
         try:
             schema_type = data['type']
         except KeyError:
-            raise ParserError("Schema does not contain 'type' property")
+            schema_type = DataType.ANY_OF
 
         try:
             data_type = DataType(schema_type)

--- a/src/openapi_parser/builders/schema.py
+++ b/src/openapi_parser/builders/schema.py
@@ -47,7 +47,7 @@ def extract_attrs(data: dict, attrs_map: Dict[str, PropertyMeta]) -> Dict[str, A
         if data.get(name) is not None
     }
 
-    attrs['type'] = DataType(attrs.get('type', DataType.ANY_OF))
+    attrs['type'] = DataType(attrs['type'])
 
     attrs.update(extract_typed_props(data, attrs_map))
 
@@ -111,6 +111,7 @@ class SchemaFactory:
         try:
             schema_type = data['type']
         except KeyError:
+            logger.warning(msg="Implicit type assignment: schema does not contain 'type' property")
             schema_type = DataType.ANY_OF
 
         try:
@@ -224,4 +225,18 @@ class SchemaFactory:
             "schemas": PropertyMeta(name="anyOf", cast=create_inner_schemas)
         }
 
-        return AnyOf(**extract_attrs(data, attrs_map))
+        if "type" in data:
+            return AnyOf(**extract_attrs(data, attrs_map))
+
+        possible_implicit_types = (
+            DataType.INTEGER, DataType.NUMBER, DataType.STRING, DataType.BOOLEAN, DataType.ARRAY, DataType.OBJECT
+        )
+        schemas = [
+            schema_factory({**data, **{"type": data_type}}) 
+            for data_type, schema_factory in self._builders.items()
+            if data_type in possible_implicit_types
+        ]
+
+        return AnyOf(type=DataType.ANY_OF, schemas=schemas)
+
+

--- a/tests/builders/schema/test_anyof.py
+++ b/tests/builders/schema/test_anyof.py
@@ -2,7 +2,7 @@ import pytest
 
 from openapi_parser.builders.schema import SchemaFactory
 from openapi_parser.enumeration import IntegerFormat
-from openapi_parser.specification import DataType, Discriminator, Integer, AnyOf, String, StringFormat
+from openapi_parser.specification import DataType, Integer, AnyOf, String, StringFormat, Number, Boolean, Array, Object
 
 data_provider = (
     (
@@ -38,6 +38,16 @@ data_provider = (
             ]
         )
     ),
+    (
+        {
+            "description": "Can be any value - string, number, boolean, array or object."
+        },
+        AnyOf(
+            type=DataType.ANY_OF,
+            description="Can be any value - string, number, boolean, array or object.",
+            schemas=[]
+        )
+    )
 )
 
 

--- a/tests/builders/schema/test_anyof.py
+++ b/tests/builders/schema/test_anyof.py
@@ -2,7 +2,10 @@ import pytest
 
 from openapi_parser.builders.schema import SchemaFactory
 from openapi_parser.enumeration import IntegerFormat
-from openapi_parser.specification import DataType, Integer, AnyOf, String, StringFormat, Number, Boolean, Array, Object
+from openapi_parser.specification import DataType, Integer, AnyOf, String, StringFormat, Array, Number, Boolean, Object, Property
+
+string_schema = String(type=DataType.STRING)
+number_schema = Number(type=DataType.NUMBER)
 
 data_provider = (
     (
@@ -40,12 +43,59 @@ data_provider = (
     ),
     (
         {
-            "description": "Can be any value - string, number, boolean, array or object."
+            "description": "Can be any type - string, number, integer, boolean, object and array"
         },
         AnyOf(
             type=DataType.ANY_OF,
-            description="Can be any value - string, number, boolean, array or object.",
-            schemas=[]
+            schemas=[
+                Integer(type=DataType.INTEGER, description="Can be any type - string, number, integer, boolean, object and array",),
+                Number(type=DataType.NUMBER, description="Can be any type - string, number, integer, boolean, object and array",),
+                String(type=DataType.STRING, description="Can be any type - string, number, integer, boolean, object and array",),
+                Boolean(type=DataType.BOOLEAN, description="Can be any type - string, number, integer, boolean, object and array",),
+                Array(type=DataType.ARRAY, description="Can be any type - string, number, integer, boolean, object and array",),
+                Object(type=DataType.OBJECT, properties=[], description="Can be any type - string, number, integer, boolean, object and array",)
+            ]
+        )
+    ),
+    (
+        {
+            "description": "Array with implicit type.",
+            "items": {"type": "string"}
+        },
+        AnyOf(
+            type=DataType.ANY_OF,
+            schemas=[
+                Integer(type=DataType.INTEGER, description="Array with implicit type."),
+                Number(type=DataType.NUMBER, description="Array with implicit type."),
+                String(type=DataType.STRING, description="Array with implicit type."),
+                Boolean(type=DataType.BOOLEAN, description="Array with implicit type."),
+                Array(type=DataType.ARRAY, items=string_schema, description="Array with implicit type."),
+                Object(type=DataType.OBJECT, properties=[], description="Array with implicit type."),
+            ]
+        )
+    ),
+    (
+        {
+            "description": "Object with implicit type.",
+            "properties": {
+                "property1": {"type": "string"},
+                "property2": {"type": "number"}
+            }
+        },
+        AnyOf(
+            type=DataType.ANY_OF,
+            schemas=[
+                Integer(type=DataType.INTEGER, description="Object with implicit type."),
+                Number(type=DataType.NUMBER, description="Object with implicit type."),
+                String(type=DataType.STRING, description="Object with implicit type."),
+                Boolean(type=DataType.BOOLEAN, description="Object with implicit type."),
+                Array(type=DataType.ARRAY, description="Object with implicit type."),
+                Object(
+                    type=DataType.OBJECT,
+                    properties=[Property("property1", string_schema), Property("property2", number_schema)],
+                    description="Object with implicit type."
+                ),
+            ]
         )
     )
 )

